### ISSUE TITLE
Performance is an issue

### ISSUE THREAD

**Issue by cooldome** [Vindaar/ggplotnim#39]
I have prepared an example where trajectories of Monte Carlo simulation have been plotted. This is a real world example, not a theoretical one. It takes 3.5 minutes to make a plot with 1000 trajectories, while R and python can cope in a matter of 2-3 seconds. 

I have made performance profiling: all time is spent in `collectScales` proc. 
I believe data is copied many many times, hence poor performance. Resulintg plot is only 140kb, so plotting on canvas takes a little

```nim
import ggplotnim
import random
import tables
import math

const paths = 1000
const dates = 80

proc gaussian*(rnd: var Rand, mu = 0.0, sigma = 1.0): float =
  var
    s = 0.0
    u = 0.0
    v = 0.0
  while s >= 1.0 or s <= 0.0:
    u = 2.0 * rnd.rand(0.0..1.0) - 1.0
    v = 2.0 * rnd.rand(0.0..1.0) - 1.0
    s = (u * u) + (v * v)

  let x = u * sqrt(-2.0 * ln(s) / s)
  return (mu + (sigma * x))

proc createDataFrame(): DataFrame = 
  const sigma = 0.10

  var rnd = initRand(124325)
  var data: OrderedTable[string, seq[Value]]
  var simPath = newSeq[Value](dates)
  for j in 0..<paths:
    simPath[0] = %~  100.0
    for i in 1..<dates:
      simPath[i] = %~(simPath[i-1].toFloat() * exp(-0.5 * sigma^2 + sigma * gaussian(rnd)))
    data["path" & $(j + 1)] = simPath

  var tenors = newSeq[Value](dates)
  for i in 0..<dates:
    tenors[i] = %~ float(i)
  data["tenors"] = tenors
  toDf(data)

let df = createDataFrame()
var g = ggplot(df)
for i in 1..paths:
  g = g + geom_line(aes(y = "path" & $i, x = "tenors"))
g + ggsave("test1.pdf")
```

**Comment by Vindaar** [comment]
Thanks a lot for the issue!
Yes, I'm aware that there's quite a few performance bugs lurking in the code base. However, the extent you mention is definitely unexpected, wow. 

I previously had some bizarre gc related performance bugs, which I could only sort of fix. I haven't checked your example yet (thanks for that!), but have you tried in what way the performance changes comparing 
- `-d:debug` and `-d:danger`
- normal `--gc:refc` and another gc, for instance boehm?

In the case I'm referring to the performance was (unexpectedly) significantly better in release mode, but not using the default gc made another big difference.

I'll take a look at it in the next few days! 


And just as another performance aside: For plots with many, many elements we currently expect much worse performance for `geom_point` than for `geom_line`, since for each cairo object we draw an individual cairo call is made, which currently includes the creation of a new cairo surface. A `geom_line` however is drawn with a single call.

**Comment by cooldome** [comment]
Yes, I did tried `-d:danger` ufortuantely I can't change `gc` in the project that easily. It affects far too many things. IMO, the main performance problem is the copies. Lot's copies of the same data being made inside `collectScales`.

**Comment by Vindaar** [comment]
> IMO, the main performance problem is the copies. Lot's copies of the same data being made inside `collectScales`.

Yeah, I agree and I'm not trying to argue otherwise. I was just curious if the same behavior showed up here too.

Just out of curiosity, what kind of project is it where you're using those Monte Carlo simulations?

**Comment by cooldome** [comment]
It is financial derivatives modelling project. Monte Carlo simulation and Partial differential equations are standard techniques in this area. Closed source I am afraid.

**Comment by Vindaar** [comment]
> It is financial derivatives modelling project. Monte Carlo simulation and Partial differential equations are standard techniques in this area. Closed source I am afraid.

No worries, I was mostly interested in what domain those Monte Carlos are used.

I'll take a look at this issue next.

**Comment by Vindaar** [comment]
Ok, I've looked at your example for a bit now. 

First of all the reason for never noticing the extremely bad performance in this case is that the whole `collectScales` code wasn't written (or tested) with many geoms in mind. In your example we create 1000 geoms, so the rather expensive code is being wayyy more often than I ever thought.
Of course it shouldn't be that slow anyways and I'll try to fix that.

Secondly, for such a plot you have in mind I would always propose to put the data frame into long format first:
```nim
  let pNames = toSeq(1 .. paths).mapIt("path" & $it)
  # convert to long table using `gather`
  df = df.gather(pNames, key = "pathNames", value = "pathValues")
```
then one can simply make a simple plotting call:
```nim
  ggplot(df, aes("tenors", "pathValues", color = "pathNames")) +
    geom_line() +
    ggsave("test1.pdf")
```
where we use the new `pathNames` column to determine the line. This brings reduces the actual runtime of the plotting code to < 1s using `-d:danger` on my machine. The drawing in your version takes a ridiculous 27 seconds. 

However, this isn't really ideal because all of the data frame manipulations are *really* unoptimized (or well, their implementation is completely naive). So the code using `gather ` is fast plotting the data frame, but takes ~ 11 s to compute the long format DF.

So for all practical purposes (for the time being anyways) I would propose to compute the long format DF directly and then plot it. Done this way the whole code runs in < 1s. As unfortunately expected almost all time is still spent drawing the plot.
Of course under certain circumstances using a long format DF might not be feasible, due to the overhead of having an essential "duplicate" column. But I guess in such a case this library would be a major bottleneck anyways, heh :|.

See the whole code below with my super unscientific "benchmarks":

```nim
import ggplotnim, sequtils
import random
import tables
import math

const paths = 1000
const dates = 80

proc gaussian*(rnd: var Rand, mu = 0.0, sigma = 1.0): float =
  var
    s = 0.0
    u = 0.0
    v = 0.0
  while s >= 1.0 or s <= 0.0:
    u = 2.0 * rnd.rand(0.0..1.0) - 1.0
    v = 2.0 * rnd.rand(0.0..1.0) - 1.0
    s = (u * u) + (v * v)

  let x = u * sqrt(-2.0 * ln(s) / s)
  return (mu + (sigma * x))

proc createDataFrame(): DataFrame =
  const sigma = 0.10

  var rnd = initRand(124325)
  var data: OrderedTable[string, seq[Value]]
  var simPath = newSeq[Value](dates)
  for j in 0..<paths:
    simPath[0] = %~  100.0
    for i in 1..<dates:
      simPath[i] = %~(simPath[i-1].toFloat() * exp(-0.5 * sigma^2 + sigma * gaussian(rnd)))
    data["path" & $(j + 1)] = simPath

  var tenors = newSeq[Value](dates)
  for i in 0..<dates:
    tenors[i] = %~ float(i)
  data["tenors"] = tenors
  toDf(data)

proc createLongDataFrame(): DataFrame =
  const sigma = 0.10
  var rnd = initRand(124325)
  var pathNames = newSeq[string](dates * paths)
  var pathVals = newSeq[float](dates * paths)
  var tenors = newSeq[int](dates * paths)
  for j in 0 ..< paths:
    pathVals[j * dates] = 100.0
    pathNames[j * dates] = "path" & $(j + 1)
    tenors[j * dates] = 0
    for i in 1..<dates:
      pathNames[j * dates + i] = "path" & $(j + 1)
      pathVals[j * dates + i] = (pathVals[j * dates + i-1] * exp(-0.5 * sigma^2 + sigma * gaussian(rnd)))
      tenors[j * dates + i] = i
  result = seqsToDf({ "tenors" : tenors,
                      "pathNames" : pathNames,
                      "pathValues" : pathVals })


import times

block:
  let t0 = epochTime()
  echo "Creating DF"
  var df = createDataFrame()
  let t1 = epochTime()
  echo "Took ", t1 - t0, "now converting to long format"
  let pNames = toSeq(1 .. paths).mapIt("path" & $it)
  # convert to long table using `gather`
  df = df.gather(pNames, key = "pathNames", value = "pathValues")
  let t2 = epochTime()
  echo "Took ", t2 - t1, "now drawing plot"
  ggplot(df, aes("tenors", "pathValues", color = "pathNames")) +
    geom_line() +
    ggsave("test1.pdf")
  let t3 = epochTime()
  echo "took ", t3 - t2

block:
  let t0 = epochTime()
  echo "Creating DF"
  let df = createLongDataFrame()
  let t1 = epochTime()
  echo "Took ", t1 - t0, " now creating plot"
  ggplot(df, aes("tenors", "pathValues", color = "pathNames")) +
    geom_line() +
    ggsave("test2.pdf")
  let t2 = epochTime()
  echo "took ", t2 - t1

```



#### tl;dr

Directly compute a long format DF and use that for much better performance. I'll see what I can do about the slowness in the use case with many geoms.

**Comment by Vindaar** [comment]
I started a rewrite of the whole data frame a few days ago. Still WIP, but I essentially reached feature parity with the default DF right now. 

Ran a mini test on the `gather` call of my code above. Timing only the `gather` call the default backend takes `~12.5 s` and the arraymancer backend... only `0.05 s`. I guess that's progress. :)

See the arraymancer backend branch and for the specific comparison I talk about:
https://github.com/Vindaar/ggplotnim/blob/arraymancerBackend/playground/testArMuchGather.nim

**Comment by cooldome** [comment]
Fantastic news!!! Thank you for your hard work

**Comment by Vindaar** [comment]
**Small update**

This slightly modified version of your original code, which uses native data types instead of `Value` on the new (WIP) arraymancer backend:

```nim
import ggplotnim
import random
import tables
import math

const paths = 1000
const dates = 80

proc gaussian*(rnd: var Rand, mu = 0.0, sigma = 1.0): float =
  var
    s = 0.0
    u = 0.0
    v = 0.0
  while s >= 1.0 or s <= 0.0:
    u = 2.0 * rnd.rand(0.0..1.0) - 1.0
    v = 2.0 * rnd.rand(0.0..1.0) - 1.0
    s = (u * u) + (v * v)

  let x = u * sqrt(-2.0 * ln(s) / s)
  return (mu + (sigma * x))

proc createDataFrame(): DataFrame =
  const sigma = 0.10
  result = initDataFrame()
  var rnd = initRand(124325)
  var simPath = newSeq[float](dates)
  for j in 0..<paths:
    simPath[0] = 100.0
    for i in 1..<dates:
      simPath[i] = simPath[i-1] * exp(-0.5 * sigma^2 + sigma * gaussian(rnd))
    result["path" & $(j + 1)] = toColumn simPath

  var tenors = newSeq[int](dates)
  for i in 0..<dates:
    tenors[i] = i
  result["tenors"] = toColumn tenors

let df = createDataFrame()
var g = ggplot(df)
for i in 1..paths:
  g = g + geom_line(aes(y = "path" & $i, x = "tenors"))
g + ggsave("test1.pdf")
```

Runs in 2 seconds flat on my machine.

Also the following example of a scatter plot with a million points:
```nim
import ggplotnim

proc main =
  var xt = randomTensor[float](1_000_000, 1.0)
  var yt = randomTensor[float](1_000_000, 1.0)

  let df = seqsToDf(xt, yt)
  ggplot(df, aes("xt", "yt")) +
    geom_point(size = some(0.1)) +
    ggsave("million.pdf")

when isMainModule:
  main()
```
runs in 16 s. However, the majority of time is spent by cairo drawing a million points as a vector graphic. If we save as a PNG the plot is created in 3.6 seconds, which seems reasonable.

While the drawing code in ggplotnim still does some things very inefficiently, the overhead at this point becomes the creation of `ginger.GraphObject` elements for each point etc., embedding the coordinates and still cairo.

As a comparison:

*matplotlib*
```python
import numpy as np
import matplotlib.pyplot as plt

def main():
  xt = np.random.rand(1_000_000)
  yt = np.random.rand(1_000_000)

  plt.plot(xt, yt, markersize = 0.1, marker = ".", fillstyle = 'full',
           color = "black", linestyle="")
  plt.savefig("million_matploltib_plot.pdf")

main()
```
takes about 22 s for me. Saving as PNG takes about 1.1 seconds.

Using `scatter` instead of plot:
```python
plt.scatter(xt, yt, s = 0.1)
```
- pdf: 24 s
- png: 2.8 s

*seaborn*
```python
import numpy as np
import seaborn

def main():
  xt = np.random.rand(1_000_000)
  yt = np.random.rand(1_000_000)

  ax = seaborn.scatterplot(xt, yt, size = 0.1)

main()
```
Seaborn is a lot slower than raw matplotlib. The above takes 136 seconds for a PDF and  21 s for a PNG. Given that seaborn is a lot more comparable to ggplotnim in terms of logic happening behind the scenes, I'd say this is pretty favorable in terms of Nim / ggplotnim.

*ggplot2*
Finally, the real competitor ( ;) ), `ggplot2`:
```R
library(ggplot2)
x <- runif(1000000, min = 0, max = 1.0)
y <- runif(1000000, min = 0, max = 1.0)
df <- data.frame(x, y)
ggplot(df, aes(x, y)) +
    geom_point(size = 0.1) +
    ggsave("million_ggplot2.pdf")
```
takes about 15.5 seconds. Unfortunately trying to save as a PNG gives me a weird font error. 


These numbers seem acceptable to me. If anyone disagrees, please let me know.

**Comment by Vindaar** [comment]
With #59 now merged, I consider this issue to be fixed. 

Please either reopen or open a new issue, if you discover performance problems again!
